File: 1570-final-prices-with-a-special-discount-in-a-shop/final-prices-with-a-special-discount-in-a-shop.cpp
class Solution {
public:
    vector<int> finalPrices(vector<int>& prices) {
        vector<int> backup(prices.size());
        vector<int> result = prices;
        stack<int> s;
        for(int i = prices.size()-1; i>=0 ; i--){
            while(!s.empty() && s.top()>prices[i]){
                int top = s.top();
                s.pop();
                backup.push_back(top);
                }

            if(!s.empty()){int disc = s.top();
            result[i] -=disc;}
            // else{}

            while(!backup.empty()){
                int elt = backup[backup.size()-1];
                backup.erase(backup.end()-1);
                s.push(elt);}
        s.push(prices[i]);
        }
        return result;
    }
};

File: 1570-final-prices-with-a-special-discount-in-a-shop/final-prices-with-a-special-discount-in-a-shop.java
class Solution {
    public int[] finalPrices(int[] prices) {
        int n = prices.length;

        int[] result = prices.clone();
        Stack<Integer> s = new Stack<>();

        for(int i=0; i<n; i++){
            while(!s.isEmpty() && prices[s.peek()] >= prices[i]){
                result[s.peek()] -= prices[i];
                s.pop();
            }
            s.add(i);
        }
        return result;
        
    }
}